### ISSUE TITLE
SDL2: Surfaces from Font with background crashes in SDL2

### ISSUE THREAD

**Issue by husano896** [pygame/pygame#742]
Some tasks to do to fix this issue.

- [ ] add unit test for problem (see code below to start) *good first issue*
- [ ] determine why no error after `image.set_colorkey((0,0,0)),`
- [ ] fix issue with 255 not properly setting alpha in SDL2
- [ ] fix alpha(127) in SDL1 (should be 50%); alpha(127) in SDL2 is working properly

---
```python
    import pygame as pg

    pg.init()

    surf = pg.display.set_mode((320,240))
    font = pg.font.SysFont("Arial", 24)
    image = font.render("Test", 0, (255,255,255), (0,0,0))
    print(image.get_colorkey()) #returns None in both SDL1/SDL2.
    image.set_alpha(255)
    surf.blit(image,(0,0))
    pg.quit()
```

When the alpha of image is set to 255 (0~254 is fine) , it raised an error:

> pygame.error: Blit combination not supported

works fine in SDL1.

But will not raise error after image.set_colorkey((0,0,0)),

Also, 
```python
    image = font.render("Test", 0, (255,255,255))
    image.set_alpha(127)
```
take no effect in SDL1 (no transparent changes) but SDL2 (shows half transparent image)

**Related Docs:** https://www.pygame.org/docs/ref/surface.html

**Comment by markph0204** [comment]
- [ ] fix issue with 255 not properly setting alpha in SDL2
- [ ] determine why no error after `image.set_colorkey((0,0,0)),`
- [ ] fix alpha(127) in SDL1 (should be 50%); alpha(127) in SDL2 is working properly

@illume can you validate this checklist?

**Comment by illume** [comment]
Thanks @markph0204 I added a check list of tasks to the top.

**Comment by nthykier** [comment]
Having dumpster dived into the SDL2 code, it appears that there are the following 3 cases where bilt can throw that error (AFAICT):

1. SDL thinks that there are less than 8 bits per pixel in the destination (`dst->format->BitsPerPixel < 8`)  
1. SDL thinks that source OR destination has more than 8 bits per pixel (`...->format->Rloss > 8`)
1. SDL fails to locate an interal blit function (not sure it can happen as it appears to always fallback to a "Slow blit" except in test cases)

(All 3 instances are in the `SDL_CalculateBlit(SDL_Surface * surface)` function - you can see it online at https://sources.debian.org/src/libsdl2/2.0.10+dfsg1-2/src/video/SDL_blit.c/?hl=290#L216 - note: link will eventually break).

**Comment by nthykier** [comment]
I think I was wrong about case 3.  It is probably the case we are hitting and the quest is now figuring out why... :-/

**Comment by nthykier** [comment]
For reference, a "quick fix" is to run `image = image.convert_alpha()` after the `image.set_alpha(255)` call.

With a bit of printf-debugging, this causes the image to get re-sized to use 4 bytes per pixel (like the surface) rather than 1 bytes per pixel.  It also seems to make "SDL_ISPIXELFORMAT_INDEXED" return false for `src->format->format` (if that rings a bell to anyone).

**Comment by nthykier** [comment]
Possibly related: https://wiki.libsdl.org/SDL_AllocFormat

"""
Returned structure may come from a shared global cache (i.e. not newly allocated), and hence should not be modified, especially the palette. Weird errors such as `Blit combination not supported` may occur. 
"""

So the question is; have we been fiddling with the SDL_PixelFormat object here (in which case it _could_ be related to #1383 as well).

**Comment by nthykier** [comment]
We are; font_render contains this gem:

```diff
diff --git a/src_c/font.c b/src_c/font.c
index 71967f60..9b3888a4 100644
--- a/src_c/font.c
+++ b/src_c/font.c
@@ -481,13 +481,6 @@ font_render(PyObject *self, PyObject *args)
     if (surf == NULL) {
         return RAISE(pgExc_SDLError, TTF_GetError());
     }
-    if (!aa && (bg_rgba_obj != NULL) && !just_return) {
-        /* turn off transparency */
-        SDL_SetColorKey(surf, 0, 0);
-        surf->format->palette->colors[0].r = backg.r;
-        surf->format->palette->colors[0].g = backg.g;
-        surf->format->palette->colors[0].b = backg.b;
-    }
     final = pgSurface_New(surf);
     if (final == NULL) {
         SDL_FreeSurface(surf);
```

Sadly, ripping it out is not sufficient for fixing the problem. :-/

**Comment by illume** [comment]
This is still an issue.

**Comment by illume** [comment]
Closed in #2213 thanks to @MyreMylar
